### PR TITLE
Fix Inefficient default GELU implementation in GPT2 #39073

### LINKED ISSUE #39073: Inefficient default GELU implementation in GPT2
While profiling the HuggingFace GPT2 model, I found that the default GELU backend used is NewGELUActivation, which is inefficient in most cases. Instead of using a fused CUDA kernel, NewGELUActivation executes multiple separate PyTorch-level operators, leading to unnecessary kernel launches and memory overhead.

```python
# activations.py:L46
class NewGELUActivation(nn.Module):
    def forward(self, input: Tensor) -> Tensor:
        return 0.5 * input * (1.0 + torch.tanh(math.sqrt(2.0 / math.pi) * (input + 0.044715 * torch.pow(input, 3.0))))
```

Is there a reason why NewGELUActivation is still used as the default for GPT2, rather than switching to nn.functional.gelu or another fused alternative?

I’d be happy to share profiler traces or help test a patch if helpful.

### PR DESCRIPTION
**Pull request by wenboqian**
# What does this PR do?

Fixes #39073 ([issue](https://github.com/huggingface/transformers/issues/39073))
### Environment
- CUDA 12.8
- Ubuntu 20.04
- torch 2.8.0
- transformers 4.55.0
## Fix
Switch `NewGELUActivation.forward` to `nn.functional.gelu(input, approximate="tanh")`.

Location: `src/transformers/activations.py`, line 47
```python
# Before
return 0.5 * input * (1.0 + torch.tanh(math.sqrt(2.0 / math.pi) * (input + 0.044715 * torch.pow(input, 3.0))))
# After
return nn.functional.gelu(input, approximate="tanh")
```

### Configuration
- Input: `torch.randn(batch_size, seq_len, hidden_dim)`
- batch_size: 16
- seq_len: 512
- hidden_dim: 768
- num_iterations: 100

### Performance
- Speedup: 7.65x (86.9%)

| Implementation | Time (ms)|
| --- | --- |
| Original (NewGELU) | 0.154 |
| Fused GELU (tanh) | 0.020 |

### Memory
- Original (NewGELU): Peak 679.47 MB, Used 96.00 MB
- Fused GELU (tanh): Peak 631.47 MB, Used 48.00 MB

### Numerical Accuracy
- Max absolute difference: 2.38e-07
- Mean absolute difference: 3.11e-10
- Mean relative error: 2.39e-09
- Conclusion: Numerically equivalent (diff < 1e-5)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by fanck0605** [comment]
I also found this performance issue. After I replaced it with nn.functional.gelu(input, approximate="tanh"), the performance improved significantly, and it doesn't affect the results on real business data. Why was this pull request closed?

According to the official documentation of Torch, these two formulas are completely equivalent.

https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.gelu.html

image
